src/utils.rs

```diff
@@ -1,40 +1,38 @@
 //! Utility functions
-use std::iter::{repeat, StepBy};
+use std::iter::StepBy;
 use std::ops::Range;
 
 #[inline(always)]
-pub fn unpack_bits<F>(buf: &mut [u8], channels: usize, bit_depth: u8, func: F)
+pub fn unpack_bits<F>(row: &[u8], buf: &mut [u8], channels: usize, bit_depth: u8, func: F)
 where
     F: Fn(u8, &mut [u8]),
 {
-    // Return early if empty. This enables to subtract `channels` later without overflow.
-    if buf.len() < channels {
-        return;
-    }
-
-    let bits = buf.len() / channels * bit_depth as usize;
-    let extra_bits = bits % 8;
-    let entries = bits / 8
-        + match extra_bits {
-            0 => 0,
-            _ => 1,
-        };
-    let skip = match extra_bits {
-        0 => 0,
-        n => (8 - n) / bit_depth as usize,
-    };
     let mask = ((1u16 << bit_depth) - 1) as u8;
-    let i = (0..entries)
-        .rev() // reverse iterator
-        .flat_map(|idx|
-            // this has to be reversed too
-            (0..8).step_by(bit_depth.into())
-            .zip(repeat(idx)))
-        .skip(skip);
-    let j = (0..=buf.len() - channels).rev().step_by(channels);
-    for ((shift, i), j) in i.zip(j) {
-        let pixel = (buf[i] & (mask << shift)) >> shift;
-        func(pixel, &mut buf[j..(j + channels)])
+
+    let mut buf_chunks = buf.chunks_exact_mut(channels);
+
+    // `shift` iterates through these ranges for each bit depth:
+    // 1 => &[7, 6, 5, 4, 3, 2, 1, 0],
+    // 2 => &[6, 4, 2, 0],
+    // 4 => &[4, 0],
+    // 8 => &[0],
+    //
+    // `(0..8).step_by(bit_depth.into()).rev()` doesn't always optimize well so
+    // shifts are calculated instead. (2023-08, Rust 1.71)
+
+    for &curr in row.iter() {
+        let mut shift = 8 - bit_depth as i32;
+
+        while shift >= 0 {
+            if let Some(chunk) = buf_chunks.next() {
+                let pixel = (curr >> shift) & mask;
+                func(pixel, chunk);
+            } else {
+                return;
```

Comment:
Slightly nervous about silently returning here. If we expect that `row` and `buf` will always have matching lengths (which I think we do?), then we should `unwrap` / `expect` instead of returning

Reply:
That makes sense.

In order to change this, I had to swap from iterating over the row in the outer loop to iterating over the chunks. When the shift cycle resets within the loop, we request the next `row` value and if it doesn't exist then it's an error.

Reply:
How about asserting that `input.len() * (8/bit_depth) * channels == output.len()` at the start of the method, and then the iteration order doesn't need to be swapped? Or is the problem that the check would fail because some of the call sites actually pass an output buffer that's larger than required?

Reply:
Yes, the checks fail. The input and output buffer sizes seem to vary enough that I can't easily derive an assertion equality that would work.

Reply:
That's weird that the checks fail, give that we intentionally set sizes for the two buffers:
https://github.com/image-rs/image-png/blob/7642f0f5a595b7b0efeb773c7279ba735f59bc0e/src/decoder/mod.rs#L582

https://github.com/image-rs/image-png/blob/7642f0f5a595b7b0efeb773c7279ba735f59bc0e/src/decoder/mod.rs#L600

Reply:
`rowlen` and `output_line_size` can be different values.

`rowlen` comes from `Reader::next_pass` and is equivalent to `self.subframe.rowlen`.

`output_line_size` is the result of `Reader::output_line_size` which calls into `ColorType::raw_row_length_from_width` with the width argument being `self.subframe.width` for non-Adam7 images.
https://github.com/image-rs/image-png/blob/7642f0f5a595b7b0efeb773c7279ba735f59bc0e/src/common.rs#L57-L69

---

Writing out a longer answer of speculation, I think I came across a solution.

We want the input buffer to have at least as many shifts as there are output entries. Then, the input row will always have enough lookups for the output buffer chunks.

```rust
let input_max_entries = input.len() * (8 / bit_depth as usize);
let output_entries = output.len() / channels;
assert!(input_max_entries >= output_entries);
```